File: csharp/beer-song/BeerSong.cs
using System.Collections.Generic;

public static class BeerSong
{
    public static string Recite(int startBottles, int takeDown) =>
        string.Join("\n\n", Verses(startBottles, takeDown)).Trim();

    private static IEnumerable<string> Verses(int startBottles, int takeDown)
    {
        for (var i = startBottles; i > startBottles - takeDown; i--)
        {
            var firstPartOfSecondLine = i switch
            {
                0 => "Go to the store and buy some more",
                1 => "Take it down and pass it around",
                _ => "Take one down and pass it around"
            };

            yield return $"{NumberOfBottles(i)} of beer on the wall, {NumberOfBottles(i).ToLower()} of beer.\n" +
                $"{firstPartOfSecondLine}, {NumberOfBottles(i - 1).ToLower()} of beer on the wall.";
        }
    }

    private static string NumberOfBottles(int numberOfBottles) => numberOfBottles switch
    {
        -1 => "99 bottles",
        0 => "No more bottles",
        1 => "1 bottle",
        _ => $"{numberOfBottles} bottles"
    };
}

File: kotlin/beer-song/src/main/kotlin/BeerSong.kt
object BeerSong {
    fun verses(startBottles: Int, takeDown: Int) =
        (startBottles downTo takeDown).joinToString("\n", transform = ::verse)

    private fun verse(bottles: Int): String {
        val howManyBottles = howManyBottles(bottles)
        val howManyBottlesLeft = howManyBottles(takeOneDown(bottles))
        val whatToDo = whatToDo(bottles)

        return "${howManyBottles.replaceFirstChar { it.uppercaseChar() }} of beer on the wall, $howManyBottles of beer.\n" +
                "${whatToDo.replaceFirstChar { it.uppercaseChar() }}, $howManyBottlesLeft of beer on the wall.\n"
    }

    private fun takeOneDown(bottles: Int) = when (bottles) {
        0 -> 99
        else -> bottles - 1
    }

    private fun howManyBottles(bottles: Int) = when (bottles) {
        0 -> "no more bottles"
        1 -> "1 bottle"
        else -> "$bottles bottles"
    }

    private fun whatToDo(bottles: Int) = when (bottles) {
        0 -> "go to the store and buy some more"
        else -> "take ${if (bottles == 1) "it" else "one"} down and pass it around"
    }
}
